helper/chksum.c

```diff
@@ -130,7 +130,7 @@ static inline int odph_process_l4_hdr(odp_packet_t      odp_pkt,
 		 * should come from the udp header, unlike for TCP where is
 		 * derived. */
 		l4_len            = odp_be_to_cpu_16(udp_hdr_ptr->length);
-		pkt_chksum_ptr    = (uint16_t *)(void *)&udp_hdr_ptr->chksum;
+		pkt_chksum_ptr    = (uint16_t *)&l4_ptr[offsetof(odph_udphdr_t, chksum)];
```

Comment:
Did you try cast to uintptr_t?

pkt_chksum_ptr = (uint16_t *)(uintptr_t)&udp_hdr_ptr->chksum;

Reply:
yes.. that does not help: 

chksum.c:133:46: warning: taking address of packed member of ‘struct <anonymous>’ may result in an unaligned pointer value [-Waddress-of-packed-member]
  133 |   pkt_chksum_ptr    = (uint16_t *)(uintptr_t)&udp_hdr_ptr->chksum;

Moreover I feel that from compiler perspective, it will treated similar to (void *) for packed structures.

Reply:
I had another thought about this. Compiler warning it because because there is no packed structure members indeed may be unaligned since compiler did not try to align it (compiler tried to put no extra padding between them and minimized the space consumed) but since base type is also not ensured to be aligned, overall we might endup having a unaligned pointer. If I align the base type also, then warning disappears: 

/** UDP header */
typedef struct **ODP_ALIGNED(4)** ODP_PACKED {
	odp_u16be_t src_port; /**< Source port */
	odp_u16be_t dst_port; /**< Destination port */
	odp_u16be_t length;   /**< UDP datagram length in bytes (header+data) */
	odp_u16be_t chksum;   /**< UDP header and data checksum (0 if not used)*/
} odph_udphdr_t;

Not sure which is better solution to address this.

Reply:
I had run into this issue as well. As a matter of fact, I saw this issue another file too, odp_classification_common.c under test/validation. The highlighted lines were:
vlan_type = (odp_u16be_t *)(void *)&ethhdr->type;
vlan_type = (uint16_t *)(void *)&vlan_hdr->type;

I found a workaround for this issue (don't know if it is correct or valid) by void casting the pointer to a local
variable before using it, like shown below:
void *ethhdr_type = (void *)&ethhdr->type;
vlan_type = (odp_u16be_t *)ethhdr_type;

Doing this for all the conflicting lines seems to suppress the warning and I don't see it during compilation anymore.

Reply:
I think a better solution would be to completely avoid taking the address by reorganizing the code a little.

Reply:
To me usage of packed attribute in protocol headers is unnecessary and should be removed. Protocol header fields are "packed" by the specifications. E.g. UDP header has four uint16_t fields and thus "packed" should not have any effect on alignment of those fields.

Currently, this warning is suppressed in odp-linux build (as GCC10 complains those).

Reply:
I agree with you @psavol. There is no need to pack udp header (struct odph_udphdr_t) since all the members are 16-bits.  odph_tcphdr_t also looked to be well packed naturally, but its using the bit-fields. 
So shall we go ahead and remove the packed attribute from both udp header and tcp header struct declarations?

Reply:
@psavol and @xvikas, we would need to consider if the above approach can probably take care of the conflicting lines in odp_classification_common.c as well. odph_vlanhdr_t doesn't need to be packed since both its members are 16-bits. As for odph_ethhdr_t, I am unsure if the above approach will work as expected as the source and destination MAC addresses are stored as array of 6 octets and alignment issues may arise.

typedef struct ODP_PACKED {
	uint8_t addr[ODPH_ETHADDR_LEN]; /**< @private Address */
} odph_ethaddr_t

typedef struct ODP_PACKED {
	odph_ethaddr_t dst; /**< Destination address */
	odph_ethaddr_t src; /**< Source address */
	odp_u16be_t type;   /**< EtherType */
} odph_ethhdr_t;

Reply:
We need the packed attribute not only for getting rid of possible padding (which in most cases would not even be added) but also for reducing the minimum alignment of the structures and their members to 1 so that we can use the corresponding struct pointer types to access memory buffers (i.e. packet data) in arbitrary, possibly unaligned, memory locations.